<a href="https://colab.research.google.com/github/RF-Lab/lab_sources/blob/master/%D0%9C%D0%B5%D1%82%D0%BE%D0%B4%D1%8B_%D0%B8_%D0%B0%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC%D1%8B_%D1%86%D0%B8%D1%84%D1%80%D0%BE%D0%B2%D0%BE%D0%B9_%D0%BE%D0%B1%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%BA%D0%B8_%D1%81%D0%B8%D0%B3%D0%BD%D0%B0%D0%BB%D0%BE%D0%B2_10_18_04_20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 13:10-14:40 БИСО-1,2-16 Метод наименьших квадратов. Остаточная ошибка.

**Лекция + Практическое занятие по дисциплине "Методы и алгоритмы цифровой обработки сигналов"**

*Кафедра КБ-3 "Управление и моделирование систем" Института Комплексной Безопасности и Специального Приборостроения, МИРЭА, 2020г.*

**Дата проведения занятия:** 18.04.2020.

**Время:** 13:10-14:40. 

**Место/Форма проведения занятия:** Дистанционное обучение/Конференция Skype. Ссылка на видео-конференцию будет предоставлена непосредственно перед занятием по почте.

**Постоянная ссылка:** https://colab.research.google.com/drive/160GFQucFiMR-egmaLBQOlww3RH9Z0uOZ

**Цель:** Анализ остаточной ошибки при оптимальной фильтрации. Разбор примеров решения задач.

**Материал прошлых лекций**:  Метод МНК позволяет решить переопределенную задачу вида:

$V\vec{c}=\vec{x}$

Путем минимизации критерия:

$Q=\left(V\vec{c}-\vec{x}\right)^T\left(V\vec{c}-\vec{x}\right)$

Такое решение можно представить в виде:

$\vec{c}=\left(V^TV\right)^{-1}V^Tx$

Решение является оптимальным в том смысле, что не существует другого набора коэффициентов $\vec{c}$, который обеспечивал бы меньшее значение критерия $Q$.

**Остаточная ошибка оптимального линейного фильтра** 

Матрица $V$ составлена из вектор-столбцов (регрессоров) - см. предыдущую лекцию

$
V=\begin{pmatrix}
a_{11} & a_{12} & \ldots & a_{1K}\\
a_{21} & a_{22} & \ldots & a_{2K}\\
\vdots & \vdots & \ldots & \vdots\\
a_{N1} & a_{N2} & \ldots & a_{NK}\\
\end{pmatrix}=
\begin{pmatrix}
\vec{a}_1 & \vec{a}_2 & \ldots \vec{a}_K \\
\end{pmatrix}
$

Здесь $\vec{a}_k$ - регрессор (другое название - терм) - вектор столбец.

Введем обозначение 

$
R = V^TV
$

$R$ - это квадратная (в отличие от $V$) матрица размером $K\times K$. Каждый элемент матрицы R можно представить в виде скалярного произведения столбцов матрицы $V$:

$r_{k,m}=\sum\limits_{n=1}^N a_{k,n} a_{m,n}$

Легко заметить, что выражение для $r_{k,m}$ является оценкой корреляции между столбцами $k$ и $m$ матрицы $V$. Если предположить, что размерность регрессоров $N$ стремится к бесконечности, то оценка коррелиции будет стремиться к истинному значению корреляции между случайными процессами выступающими в качестве регрессоров. Будем называть матрицу $R$ **корреляционной матрицей**. 

Элементы корреляционной матрицы обладают следующими свойствами (*приведите доказательнство*):

* $r_{k,m}=r_{m,k}$ - для действительнозначных регрессоров.
* $r_{k,k}\ge 0$

Также можно доказать, что корреляционная матрица является симметричной неотрицательно определенной матрицей. 
Т.е. $\forall\vec{y}$ билинейная форма вида неотрицательна:

$\vec{y}^TR\vec{y}\ge 0$

Это свойство позволяет сократить объем вычислений при поиске $R^{-1}$ (можно использовать разложение Холесского, см.numpy.linalg.cholesky(a)).

Введем обозначение

$\vec{\rho}=V^Tx$

Каждый элемент вектора $\vec{\rho}$ можно представить в виде:

$\rho_k=\sum\limits_{n=1}^Na_{n,k}x_n$

Т.е. $\rho_k$ - определяет оценку корреляции между $k$-ым вектором-регрессором $\vec{a}_k$ и вектором требуемого выхода $\vec{x}$.

В таких обозначениях решение переопределенной системы линейных уравнений методом наименьших квадратов можно представить в виде:

\begin{equation}
\vec{c}=\left(V^TV\right)^{-1}V^Tx=R^{-1}\vec{\rho}\tag{1}
\end{equation}

Уравнение $(\ref{1})$ в форме:
\begin{equation}
R\vec{c}=\vec{\rho}\tag{2}
\end{equation}
называется *уравнением Винера*.

Рассмотрим критерий оптимизации - сумму квадратов отклонений:

$
Q=\left(V\vec{c}-\vec{x}\right)^T\left(V\vec{c}-\vec{x}\right)=
(\vec{c}^TV^T-\vec{x}^T)\left(V\vec{c}-\vec{x}\right)=\\
\vec{c}^TV^TV\vec{c}-\vec{c}^TV^T\vec{x}-\vec{x}^TV\vec{c}+\vec{x}^T\vec{x}
$

Учитывая, что для вычислений в действительных числах $\vec{c}^TV^T\vec{x}=\vec{x}^TV\vec{c}$ (*докажите*):

\begin{equation}
Q(\vec{c})=\vec{c}^TV^TV\vec{c}-2\vec{c}^TV^T\vec{x}+\vec{x}^T\vec{x}=
\vec{x}^T\vec{x}-2\vec{c}^T\vec{\rho}+\vec{c}^TR\vec{c}\tag{3}
\end{equation}

Подставим уравнение Винера $R\vec{c}=\vec{\rho}$ в выражение $(\ref{3})$:

\begin{equation}
Q_{opt}=\vec{x}^T\vec{x}-2\vec{c}^T\vec{\rho}+\vec{c}^T\vec{\rho}=\\
\vec{x}^T\vec{x}-\vec{c}^T\vec{\rho}=\\
\vec{x}^T\vec{x}-\vec{c}^TR\vec{c}\tag{3}
\end{equation}

* Остаточная ошибка оптимального фильтра $Q_{opt}$ - это квадрат модуля кратчайшего расстояния от линейной оболочки, образованной столбцами (регрессорами) матрицы $V$ до точки $\vec{x}$.

* Остаточную ошибку оптимального фильтра $Q_{opt}$ нельзя уменьшить в рамках заданного набора регрессоров и вектора $\vec{x}$.

* Выражение $\vec{c}^TR\vec{c}\ge 0$, поэтому значение ошибки лежит в диапазоне $0\ldots \vec{x}^T\vec{x}$.

Расчитаем остаточную ошибку для задачи 1 из предыдущей лекции:


In [33]:
import numpy as np
import matplotlib.pyplot as plt

N     = 10000
alpha = 2
beta  = -3
x     = np.array([np.cos( np.pi/N*4*np.arange(0.,N) )+np.cos( np.pi/N*20*np.arange(0.,N) )+np.cos( np.pi/N*30*np.arange(0.,N) )]).transpose()
eta   = 0.9 * np.random.randn( N,1 )
xi    = 2.1 * np.random.randn( N,1 )
a     = alpha*x + eta
b     = beta*x + xi

V     = np.hstack((a,b))
R     = V.transpose()@V
rho   = V.transpose()@x
c     = np.linalg.inv(R)@rho
print('c={0}'.format(c))

sol_x = V@c

c=[[ 0.32158483]
 [-0.09020791]]


Сначала раcсчитаем ошибку напрямую $(\hat{\vec{x}}-\vec{x})^T(\hat{\vec{x}}-\vec{x})=\sum\limits_{n=1}^N(\hat{x}_n-x_n)^2$

In [6]:
Q = (sol_x-x).transpose()@(sol_x-x)
print( 'Q={0}\n'.format(Q) )

Q=[[1345.85331338]]



Теперь будем использовать формулу $Q_{opt}=\vec{x}^T\vec{x}-\vec{c}^TR\vec{c}$

In [7]:
Q = x.transpose()@x - c.transpose()@R@c
print( 'Q={0}\n'.format(Q) )

Q=[[1345.85331338]]



Максимальная ошибка для данного примера:

In [8]:
Qmax = x.transpose()@x
print( 'Qmax={0}\n'.format(Qmax) )

Qmax=[[15000.]]



Разделив сумму квадратов ошибок на максимальное теоретическое значение суммы квадратов отклонений, получим относительную ошибку:

In [14]:
print('Относительная ошибка {0}'.format(Q/Qmax))

Относительная ошибка [[0.08972355]]


В цифровой обработке традиционно используются логарифмические величины для относительной ошибки:

In [23]:
nmseDb = 10*np.log10(Q/Qmax)
print('Относительная ошибка {0:5.2f}dB'.format(nmseDb[0][0]))

Относительная ошибка -10.47dB


Сравним с относительной ошбкой во входных векторах $\vec{a}$ и $\vec{b}$:

In [36]:
nmse_a      = (a-x*alpha).transpose()@(a-x*alpha)
nmse_b      = (b-x*beta).transpose()@(b-x*beta)
rms_x       = x.transpose()@x
nmse_a_db   = 10*np.log10(nmse_a[0][0]/rms_x[0,0]/(alpha*alpha))
nmse_b_db   = 10*np.log10(nmse_b[0][0]/rms_x[0,0]/(beta*beta))
print( 'Относительные ошибки входных данных a={:5.2f}dB b={:5.2f}dB'.format( nmse_a_db, nmse_b_db ) )

Относительные ошибки входных данных a=-8.65dB b=-4.84dB


Таким образом, относительная ошибка
* по сравнению со входным вектором $\vec{a}$ улучшилась на -8.65-(-10.47)=1.82дБ, т.е. в $10^{1.82/10}=1.52$ раза.
* по сравнению со входным вектором $\vec{b}$ улучшилась на -4.84-(-10.47)=5.63дБ, т.е. в $10^{5.63/10}=3.65$ раза.

Рассмотрим пример с очисткой звукового файла. Нам понадобится запись с "чистым" звуком (вектор $\vec{x}$). 

In [38]:
from IPython.display import Audio
x = np.loadtxt("https://raw.githubusercontent.com/RF-Lab/lab_sources/master/x_sound_16000.txt")
Audio(x,rate=16000)

Добавим к чистому сигналу шум - получим вектор 
$\vec{a}=\vec{x}+\vec{\eta}$

In [43]:
a = x + 500*np.random.randn( *x.shape )
Audio(a,rate=16000)